[compiler] Expose relay ops registered as torch.ops.tvm.__

bwasti opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #44 [compiler] Expose `to_relay` API which hands the Relay graph over in Python
* **#43 [compiler] Expose relay ops registered as torch.ops.tvm.__**

bwasti committed:
[compiler] Expose relay ops registered as torch.ops.tvm.__

bwasti commented:
There is currently a bug lurking here regarding the lifetime management of DLManagedTensors.  TVM stores NDArrays in the graph runtime and on program exit will attempt to delete them, assuming that it has ownership of the DLManagedTensors' tensors.  This ends up breaking everything and the program segfaults.  A small patch to get around this is

```
diff --git a/src/runtime/ndarray.cc b/src/runtime/ndarray.cc
index 39c17b8b..92747b1b 100644
--- a/src/runtime/ndarray.cc
+++ b/src/runtime/ndarray.cc
@@ -190,8 +190,10 @@ void NDArray::CopyFromTo(DLTensor* from,
 using namespace tvm::runtime;

 void NDArrayDLPackDeleter(DLManagedTensor* tensor) {
-  static_cast<NDArray::Container*>(tensor->manager_ctx)->DecRef();
-  delete tensor;
+  if (tensor) {
+    static_cast<NDArray::Container*>(tensor->manager_ctx)->DecRef();
+    delete tensor;
+  }
 }

 int TVMArrayAlloc(const tvm_index_t* shape,
```

but realistically we should revisit the ownership model and how values are passed in to TVM

bwasti committed:
Update on "[compiler] Expose relay ops registered as torch.ops.tvm.__"

[compiler] Expose relay ops registered as torch.ops.tvm.__

gh-metadata: pytorch tvm 43 gh/bwasti/27/head

bwasti committed:
Update on "[compiler] Expose relay ops registered as torch.ops.tvm.__"

[compiler] Expose relay ops registered as torch.ops.tvm.__

gh-metadata: pytorch tvm 43 gh/bwasti/27/head

bwasti committed:
Update on "[compiler] Expose relay ops registered as torch.ops.tvm.__"

[compiler] Expose relay ops registered as torch.ops.tvm.__

gh-metadata: pytorch tvm 43 gh/bwasti/27/head

bwasti committed:
Update on "[compiler] Expose relay ops registered as torch.ops.tvm.__"

[compiler] Expose relay ops registered as torch.ops.tvm.__

gh-metadata: pytorch tvm 43 gh/bwasti/27/head

bwasti committed:
Update on "[compiler] Expose relay ops registered as torch.ops.tvm.__"

[compiler] Expose relay ops registered as torch.ops.tvm.__

gh-metadata: pytorch tvm 43 gh/bwasti/27/head

wanchaol commented on torch_tvm/operators.cpp:
```diff
@@ -14,12 +20,50 @@ std::unordered_map<Symbol, TVMOpFunctor>& getTVMOperatorMap() {
+        Node* node =
+            wrapper_graph.create(op.sym, torch_inputs, schema.returns().size());
+        wrapper_graph.appendNode(node);
+        wrapper_graph.registerOutput(node->output());
```
this is a bit confusing. `wrapper_graph` create and append the node into its graph, then it does not ever get used when we creating the `SingletonSubgraph`?

bwasti replied:
it's a lifetime management hack

wanchaol commented on torch_tvm/operators.cpp:
```diff
@@ -14,12 +20,50 @@ std::unordered_map<Symbol, TVMOpFunctor>& getTVMOperatorMap() {
+
+        // NB: We assume all relay ops are pure
+        auto options = OperatorOptions().aliasAnalysis(AliasAnalysisKind::PURE);
+        // We don't know the true number of inputs so we make it varargs
```
we do know the true number of arguments right? even if in runtime, the number of inputs are the same as `schema.arguments()` I believe.

bwasti replied:
oh yea we do! dated comment

bwasti commented on torch_tvm/compiler.cpp:
```diff
@@ -214,7 +216,7 @@ void TVMCompiler::run(Stack& stack) {
     tvm::relay::Function tvm_func;
     try {
-      tvm_func = convertToRelay(subgraph_, &input_values_);
+      tvm_func = convertToRelay(subgraph_, &cache_[spec].input_values);
```
C++ unordered_map's [] creates an object if it doesn't exist

bwasti committed:
Update on "[compiler] Expose relay ops registered as torch.ops.tvm.__"

[compiler] Expose relay ops registered as torch.ops.tvm.__

gh-metadata: pytorch tvm 43 gh/bwasti/27/head

wanchaol reviewed (approved):
looks good
